Tuple Support

fordN opened the pull request:
This PR adds support for`Tuple` parameter types  (`structs` in Solidity and `tuples` in ABI.)

A `Tuple` has similarities to fixed and dynamic arrays, but can contain a combination of different `ParamTypes` and depending on the types it contains will it will be encoded differently:
  - If there are only static types in the tuple, they are padded to 32 bytes and put one after the other.
  - If there are any dynamic types in the tuple the entire tuple is considered dynamic and prefixed with an offset to the tuple location.  Each of it's parameters are encoded subsequently with their head representing an offset where you can find the encoded element, if an element is static the head is the encoded element. 

One can access the `tuple` encoding and decoding functionality on the command line through `ethabi-cli`:
  - syntax: `(type,type,type...)`
  - examples: 
        ```
         ethabi encode params -v '(bool,bool,bool,bool,string)''([0,1,1,0,hello)'
        ```
        ```
         ethabi decode params -t '(bool,bool,bool,bool,string])' 0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000568656c6c6f000000000000000000000000000000000000000000000000000000
        ```

Also included in the PR are minor updates to mend some FixedArray tests and to align the FixedArray and Tuple implementations since they have strong similarities. 

_Note: This PR does not include a derive implementation for `ParamType::Tuple`_

dvdplm commented on ethabi/src/decoder.rs:
```diff
@@ -34,6 +34,19 @@ fn as_u32(slice: &Word) -> Result<u32, Error> {
+		return Err(Error::InvalidData);
+	}
+
+	let result = ((slice[28] as usize) << 24) +
```
I'm not sure I understand the purpose of `as_usize()` tbh. It seems to be the same code as `as_u32()` but truncating 8 bytes to 4? Is it there as a convenience function to get an index for slicing?

fordN replied:
@dvdplm exactly, it's for convenience and code readability.  
Slice indices must be `usize`. As I was optimizing to remove unnecessary assignments and make it more direct, it became hard to read with the `as usize`  statements in the middle of a line of code.    

I see your point though, @ordian, I haven't tested across platforms and as a convenience function it may be better to convert the final item instead of the individual slices to usize.

ordian replied:
@fordN your code if correct on for both 32 and 64-bit platforms, I'm saying when reading code it something to be aware of as `std::mem::size_of::<usize>()` will differ for `wasm32` and `x86_64` targets. That's why I actually prefer it to be explicit, but for convenience I would recommend wrapping it as `as_u32(...).map(|i| usize::try_from(i).expect("usize is at least 32-bits"))` to avoid code duplication and use it everywhere for consistency.

fordN replied:
@ordian Thanks for explaining further, I see your point.
I've updated it to simply used the `as usize` conversion  as it was before and as it is elsewhere in the project; it doesn't "muddy up" the code as much as I thought it might.

dvdplm commented on rustfmt.toml:
```diff
@@ -0,0 +1,2 @@
+hard_tabs = true
+use_small_heuristics = 'Off'
```
We might want to use what we use in `parity-common`:

```
hard_tabs = true
max_width = 120
use_small_heuristics = "Max"
edition = "2018"
```

fordN replied:
I've updated `rustfmt.toml`, but did not reformat the files in the project with it since I want to keep the changes in this PR clean and easy to review. I'd suggest running `cargo fmt` after this PR is merged and also including a format check in the travis build (see below) to ensure consistent formatting going forward. 
Let me know how you want to proceed, I'm happy to submit another commit or a separate PR with reformatting of the whole project and a travis build update. 

Travis build step to check formatting:
```yml
  - env: CHECK_FORMATTING=true
    rust: stable
    script:
      - rustup component add rustfmt
      - cargo fmt --all -- --check
  ```

dvdplm replied:
You are right, let's do it in a separate PR. :)

dvdplm commented:
So at this point outstanding issues are:

1. the question of manual implementations of `Deserialize`
1. an issue for the missing `derive` impls
1. the `as_usize()`

Did I miss anything?

fordN commented:
Ok, getting there. :) 

Regarding the outstanding issues you mentioned: 
- [ ]  I implemented the deserialization manually because I wasn't able to get it right using derived `Deserialize`, but I'll do another quick pass at it this afternoon.
- [X]   I've created an issue for the derive impls, you can find it [here](https://github.com/paritytech/ethabi/issues/175).
- [X]   I removed the `as_usize()` helper function in favor of the more explicit paradigm that has been used in the past: `u32() as usize`.

dvdplm commented:
>     * I implemented the deserialization manually because I wasn't able to get it right using derived `Deserialize`, but I'll do another quick pass at it this afternoon.

Cool, thank you for the effort, really appreciated. Let us know how it goes and if it can't be done we'll merge then. :)

fordN commented:
I made an attempt yesterday at removing the custom deserializer impls in favor of deriving the deserialization, but to no avail.  It's tricky because `ParamType::Tuple` is adjacently tagged, but it is the only variant with that behavior. Also in the case of `ParamType::Tuple` the parent `EventParam` has a different structure, hence the `TupleParam` type that I created for use during deserialization.  

To make it work with derived deserializers: `EventParam` could be updated to be an enum with the current `EventParam` and `TupleParam` being the variants and/or `ParamType::Tuple` could be treated differently by applying a deserializer function to it using the `deserialize_with` field attribute.  Continued attention to this PR is getting out of scope for me though, so if you want to explore these options I'd appreciate if we could save it for a subsequent PR to which I'd be happy to provide input. :)
